x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datatiers/NodesDataTiersUsageTransportAction.java

```diff
@@ -0,0 +1,241 @@
+        Map<String, NodeDataTiersUsage.UsageStats> usageStatsByTier = aggregateStats(routingNode, state.metadata(), nodeIndicesStats);
+        return new NodeDataTiersUsage(clusterService.localNode(), usageStatsByTier);
+    }
+
+    // For testing
+    static Map<String, NodeDataTiersUsage.UsageStats> aggregateStats(
+        RoutingNode routingNode,
+        Metadata metadata,
+        NodeIndicesStats nodeIndicesStats
+    ) {
+        if (routingNode == null) {
+            return Map.of();
+        }
+        Map<String, NodeDataTiersUsage.UsageStats> usageStatsByTier = new HashMap<>();
+        Set<String> localIndices = StreamSupport.stream(routingNode.spliterator(), false)
+            .map(routing -> routing.index().getName())
+            .collect(Collectors.toSet());
+        for (String indexName : localIndices) {
+            IndexMetadata indexMetadata = metadata.index(indexName);
+            String tier = findPreferredTier(indexMetadata);
+            if (tier != null) {
+                NodeDataTiersUsage.UsageStats usageStats = usageStatsByTier.computeIfAbsent(
+                    tier,
+                    ignored -> new NodeDataTiersUsage.UsageStats()
+                );
+                usageStats.addIndex(indexName);
+                List<IndexShardStats> allShardStats = nodeIndicesStats.getShardStats(indexMetadata.getIndex());
+                if (allShardStats != null) {
+                    for (IndexShardStats indexShardStats : allShardStats) {
+                        usageStats.incrementTotalSize(indexShardStats.getTotal().getStore().totalDataSetSizeInBytes());
+                        usageStats.incrementDocCount(indexShardStats.getTotal().getDocs().getCount());
+
+                        ShardRouting shardRouting = routingNode.getByShardId(indexShardStats.getShardId());
+                        if (shardRouting != null && shardRouting.state() == ShardRoutingState.STARTED) {
+                            usageStats.incrementTotalShardCount(1);
+
+                            // Accumulate stats about started primary shards
+                            StoreStats primaryStoreStats = indexShardStats.getPrimary().getStore();
+                            if (shardRouting.primary() && primaryStoreStats != null) {
+                                usageStats.addPrimaryShardSize(primaryStoreStats.totalDataSetSizeInBytes());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return usageStatsByTier;
+    }
+
+    @Nullable
+    static String findPreferredTier(IndexMetadata indexMetadata) {
+        List<String> tierPref = DataTier.parseTierList(indexMetadata.getSettings().get(DataTier.TIER_PREFERENCE))
+            .stream()
+            .filter(DataTier::validTierName)
+            .toList();
+        if (tierPref.isEmpty() == false) {
+            return tierPref.get(0);
+        }
```

Comment:
I think we can use `IndexMetadata#tierPreference` here to avoid reading the setting and parsing it for every index.